File: containertrack-frontend/src/types/container.ts
export type ContainerStatus =
  | "REGISTERED"
  | "DEPARTED_ORIGIN"
  | "ARRIVED_PORT"
  | "DEPARTED_PORT"
  | "ARRIVED_WAREHOUSE"
  | "DISCHARGED";

/** Pseudo-status used only in the UI to render the delay flag as an overlay badge. */
export const DELAY_FLAG_STATUS = "DELAY_FLAG" as const;

export const CONTAINER_STATUS_ORDER: ContainerStatus[] = [
  "REGISTERED",
  "DEPARTED_ORIGIN",
  "ARRIVED_PORT",
  "DEPARTED_PORT",
  "ARRIVED_WAREHOUSE",
  "DISCHARGED",
];

export interface Container {
  id: string;
  containerNumber: string;
  blNumber: string;
  // Backend ContainerDTO is flat (shippingCompanyId/shippingCompanyName etc), not
  // nested objects — matches the real wire contract, verified against the live API.
  shippingCompanyId: string;
  shippingCompanyName: string;
  landCarrierId: string | null;
  landCarrierName: string | null;
  warehouseAssigneeId: string | null;
  warehouseAssigneeName: string | null;
  originPort: string;
  destinationPort: string;
  cargoDescription: string;
  responsibleOperatorId: string;
  responsibleOperatorName: string;
  status: ContainerStatus;
  version: number;
  lastUpdatedBy: string | null;
  lastUpdatedByName: string | null;
  lastUpdatedAt: string | null;
  estimatedDepartureDate: string | null;
  actualDepartureDate: string | null;
  estimatedArrivalPort: string | null;
  actualArrivalPort: string | null;
  freeDaysLimit: number | null;
  freeDaysExpiry: string | null;
  estimatedDeparturePort: string | null;
  actualDeparturePort: string | null;
  estimatedArrivalWarehouse: string | null;
  actualArrivalWarehouse: string | null;
  dischargeStartAt: string | null;
  dischargeEndAt: string | null;
  dischargeNotes: string | null;
  delayFlag: boolean;
  internalNotes: string | null;
  createdAt: string;
}

export interface ContainerListParams {
  page?: number;
  size?: number;
  status?: ContainerStatus;
  shippingCompanyId?: string;
  dateFrom?: string;
  dateTo?: string;
  operatorId?: string;
  sort?: string;
}

export interface CreateContainerRequest {
  containerNumber: string;
  blNumber: string;
  // Backend expects these as Long (JSON number). React-hook-form/<select> values
  // are always strings — convert with Number(...) at the call site before sending.
  shippingCompanyId: number;
  // Optional at creation — filled in once the container reaches port.
  landCarrierId?: number;
  originPort: string;
  destinationPort: string;
  cargoDescription: string;
  responsibleOperatorId: number;
  estimatedDepartureDate: string;
  internalNotes?: string;
}

export type UpdateContainerRequest = Partial<
  Omit<CreateContainerRequest, "containerNumber" | "shippingCompanyId" | "responsibleOperatorId" | "landCarrierId">
> & {
  shippingCompanyId?: number;
  landCarrierId?: number;
  responsibleOperatorId?: number;
  version: number;
  // Additional editable estimate fields, not part of container creation but
  // patchable afterwards (e.g. from the calendar's inline date-edit modal).
  estimatedArrivalWarehouse?: string;
  freeDaysExpiry?: string;
};

export interface TransitionRequest {
  targetStatus: ContainerStatus;
  actualDepartureDate?: string;
  // Adjustable per CU-05.1 when registering departure from origin.
  estimatedArrivalPort?: string;
  actualArrivalPort?: string;
  // Field name must match backend exactly (TransitionRequest.freeDaysLimitOverride) —
  // sending "freeDaysLimit" here is silently dropped by Jackson (unknown property),
  // so the value the user enters never actually applies.
  freeDaysLimitOverride?: number;
  actualDeparturePort?: string;
  estimatedArrivalWarehouse?: string;
  actualArrivalWarehouse?: string;
}

// Backend GET /containers/{id}/history returns a single flat, chronologically-sorted
// array — not a {fieldChanges, statusChanges} wrapper object — with a `type` discriminant
// per entry. Matches the real FieldChangeDTO contract, verified against the live API.
export interface HistoryEntry {
  type: "FIELD_CHANGE" | "STATUS_CHANGE";
  fieldName: string;
  oldValue: string | null;
  newValue: string | null;
  updatedById: string;
  updatedByName: string | null;
  updatedAt: string;
}

export type ContainerHistory = HistoryEntry[];

export interface UploadedPhoto {
  photoId: string;
  r2Key: string;
  presignedUrl: string;
}

export interface PhotoUploadFailure {
  filename: string;
  reason: string;
}

export interface PhotoUploadResponse {
  uploaded: UploadedPhoto[];
  failed: PhotoUploadFailure[];
}

export interface ContainerPhoto {
  photoId: string;
  presignedUrl: string;
  originalFilename: string;
  uploadedAt: string;
  // Backend PhotoDTO is flat (uploadedById/uploadedByName), not a nested object.
  uploadedById: string | null;
  uploadedByName: string | null;
  isValid: boolean;
}

export interface DischargeRequest {
  dischargeStartAt: string;
  dischargeEndAt: string;
  dischargeNotes: string;
}

export interface ContainerRealtimeMessage {
  containerId: string;
  changedFields: { fieldName: string; oldValue: string | null; newValue: string | null }[];
  updatedBy: string;
  updatedByName: string;
  updatedAt: string;
}

File: backend/src/main/java/com/containertrack/entity/Container.java
package com.containertrack.entity;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.time.OffsetDateTime;

@Entity
@Table(name = "containers")
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Container {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "container_number", nullable = false, length = 11)
    private String containerNumber;

    @Column(name = "bl_number", nullable = false, length = 50)
    private String blNumber;

    @Column(name = "shipping_company_id", nullable = false)
    private Long shippingCompanyId;

    @Column(name = "land_carrier_id")
    private Long landCarrierId;

    @Column(name = "warehouse_assignee_id")
    private Long warehouseAssigneeId;

    @Column(name = "origin_port", nullable = false)
    private String originPort;

    @Column(name = "destination_port", nullable = false)
    private String destinationPort;

    @Column(name = "cargo_description", length = 500)
    private String cargoDescription;

    @Column(name = "responsible_operator_id", nullable = false)
    private Long responsibleOperatorId;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 30)
    @Builder.Default
    private ContainerStatus status = ContainerStatus.REGISTERED;

    @Version
    @Column(nullable = false)
    private Integer version;

    @Column(name = "last_updated_by")
    private Long lastUpdatedBy;

    @Column(name = "last_updated_at")
    private OffsetDateTime lastUpdatedAt;

    @Column(name = "estimated_departure_date", nullable = false)
    private OffsetDateTime estimatedDepartureDate;

    @Column(name = "actual_departure_date")
    private OffsetDateTime actualDepartureDate;

    @Column(name = "estimated_arrival_port")
    private OffsetDateTime estimatedArrivalPort;

    @Column(name = "actual_arrival_port")
    private OffsetDateTime actualArrivalPort;

    @Column(name = "free_days_limit", nullable = false)
    @Builder.Default
    private Integer freeDaysLimit = 10;

    @Column(name = "free_days_expiry")
    private OffsetDateTime freeDaysExpiry;

    @Column(name = "estimated_departure_port")
    private OffsetDateTime estimatedDeparturePort;

    @Column(name = "actual_departure_port")
    private OffsetDateTime actualDeparturePort;

    @Column(name = "estimated_arrival_warehouse")
    private OffsetDateTime estimatedArrivalWarehouse;

    @Column(name = "actual_arrival_warehouse")
    private OffsetDateTime actualArrivalWarehouse;

    @Column(name = "discharge_start_at")
    private OffsetDateTime dischargeStartAt;

    @Column(name = "discharge_end_at")
    private OffsetDateTime dischargeEndAt;

    @Column(name = "discharge_notes", columnDefinition = "TEXT")
    private String dischargeNotes;

    @Column(name = "delay_flag", nullable = false)
    @Builder.Default
    private Boolean delayFlag = false;

    @Column(name = "internal_notes", columnDefinition = "TEXT")
    private String internalNotes;

    @Column(name = "created_by", nullable = false)
    private Long createdBy;

    @Column(name = "deleted_at")
    private OffsetDateTime deletedAt;

    @CreationTimestamp
    @Column(name = "created_at", nullable = false, updatable = false)
    private OffsetDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at", nullable = false)
    private OffsetDateTime updatedAt;
}
